src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseMessage.cs

```diff
@@ -181,19 +183,62 @@ private static byte[] DecodeSignature(CborReader reader)
             return reader.ReadByteString();
         }
 
-        internal static int CreateToBeSigned(string context, ReadOnlySpan<byte> encodedProtectedHeader, ReadOnlySpan<byte> content, Span<byte> destination)
+        internal static int CreateHashedToBeSigned(Span<byte> destination, string context, ReadOnlySpan<byte> encodedProtectedHeader, ReadOnlySpan<byte> content, HashAlgorithmName hashAlgorithm)
+        {
+            int bytesWritten = CreateToBeSigned(destination, context, encodedProtectedHeader, ReadOnlySpan<byte>.Empty);
+            bytesWritten -= 1; // Trim the empty bstr content, it is just a placeholder.
+
+            using (var hasher = IncrementalHash.CreateHash(hashAlgorithm))
+            {
+                hasher.AppendData(destination.Slice(0, bytesWritten));
+
+                // content length
+                CoseHelpers.WriteByteStringLength(hasher, (ulong)content.Length);
+
+                //content
+                hasher.AppendData(content);
+
+                return hasher.GetHashAndReset(destination);
+            }
+        }
+
+        internal static async Task<int> CreateHashedToBeSignedAsync(byte[] destination, string context, ReadOnlyMemory<byte> encodedProtectedHeader, Stream content, HashAlgorithmName hashAlgorithm)
+        {
+            int bytesWritten = CreateToBeSigned(destination, context, encodedProtectedHeader.Span, ReadOnlySpan<byte>.Empty);
+            bytesWritten -= 1; // Trim the empty bstr content, it is just a placeholder.
+
+            using (IncrementalHash hasher = IncrementalHash.CreateHash(hashAlgorithm))
+            {
+                hasher.AppendData(destination, 0, bytesWritten);
+
+                //content length
+                CoseHelpers.WriteByteStringLength(hasher, (ulong)(content.Length - content.Position));
+
+                // content
+                byte[] buffer = new byte[4096];
+                int bytesRead;
+#pragma warning disable CA1835 // Prefer the 'Memory'-based overloads for 'ReadAsync' and 'WriteAsync'
+                while ((bytesRead = await content.ReadAsync(buffer, 0, buffer.Length).ConfigureAwait(false)) > 0)
+#pragma warning restore CA1835
+                {
+                    hasher.AppendData(buffer, 0, bytesRead);
+                }
+
+                return hasher.GetHashAndReset(destination);
```

Comment:
I like what you've done here conceptually.  But since .NET Framework has to turn this span back into an array, it might be better to return the IncrementalHash object (or just receive it as input) so that when it comes time to sign we can do something like

```C#
#if NETFRAMEWORK
    byte[] hash = hasher.GetHashAndReset();
    return key.SignHash(hash, ...);
#else
    Debug.Assert(hasher.HashLengthInBytes <= 512 / 8);
    Span<byte> hash = stackalloc byte[hasher.HashLengthInBytes];
    hasher.GetHashAndReset(hash);
    return key.SignHash(hash, ...);
#endif